overlord/snapstate/target.go

```diff
@@ -754,3 +818,471 @@ func validatedComponentInfo(path string, si *snap.Info, csi *snap.ComponentSideI
+}
+
+func validateAndFilterPlanRefreshes(st *state.State, plan *updatePlan, opts Options) error {
+	if ValidateRefreshes == nil || len(plan.targets) == 0 || opts.Flags.IgnoreValidation {
+		return nil
+	}
+
+	ignoreValidation := make(map[string]bool, len(plan.targets))
+	for _, t := range plan.targets {
+		if t.snapst.IgnoreValidation {
+			ignoreValidation[t.info.InstanceName()] = true
+		}
+	}
+
+	validated, err := ValidateRefreshes(st, plan.targetInfos(), ignoreValidation, opts.UserID, opts.DeviceCtx)
+	if err != nil {
+		if !plan.refreshAll() {
+			return err
+		}
+		logger.Noticef("cannot refresh some snaps: %v", err)
+	}
+
+	validatedMap := make(map[string]bool, len(validated))
+	for _, sn := range validated {
+		validatedMap[sn.InstanceName()] = true
+	}
+
+	plan.filter(func(t target) (bool, error) {
+		_, ok := validatedMap[t.info.InstanceName()]
+		return ok, nil
+	})
+
+	return nil
+}
+
+func filterHeldSnapsInPlan(st *state.State, plan *updatePlan, isAutoRefresh bool) error {
+	holdLevel := HoldGeneral
+	if isAutoRefresh {
+		holdLevel = HoldAutoRefresh
+	}
+
+	heldSnaps, err := HeldSnaps(st, holdLevel)
+	if err != nil {
+		return err
+	}
+
+	plan.filter(func(t target) (bool, error) {
+		_, ok := heldSnaps[t.info.InstanceName()]
+		return !ok, nil
+	})
+
+	return nil
+}
+
+// storeInstallGoal implements the UpdateGoal interface and represents a group
+// of snaps that are to be updated from the store.
+type storeUpdateGoal struct {
+	// snaps is a mapping of snap names to StoreUpdate structs.
+	snaps map[string]StoreUpdate
+}
+
+// StoreUpdate represents a snap that is to be updated from the store.
+type StoreUpdate struct {
+	// InstanceName is the instancename of snap to update.
+	InstanceName string
+	// RevOpts contains options that apply to the update of this snap.
+	RevOpts RevisionOptions
+}
+
+// StoreUpdateGoal creates a new UpdateGoal to update snaps from the store.
+func StoreUpdateGoal(snaps ...StoreUpdate) UpdateGoal {
+	mapping := make(map[string]StoreUpdate, len(snaps))
+	for _, sn := range snaps {
+		if _, ok := mapping[sn.InstanceName]; ok {
+			continue
+		}
+
+		mapping[sn.InstanceName] = sn
+	}
+
+	return &storeUpdateGoal{
+		snaps: mapping,
+	}
+}
+
+func (s *storeUpdateGoal) toUpdate(ctx context.Context, st *state.State, opts Options) (updatePlan, error) {
+	if opts.ExpectOneSnap && len(s.snaps) != 1 {
+		return updatePlan{}, ErrExpectedOneSnap
+	}
+
+	if err := validateAndInitStoreUpdates(st, s.snaps, opts); err != nil {
+		return updatePlan{}, err
+	}
+
+	user, err := userFromUserID(st, opts.UserID)
+	if err != nil {
+		return updatePlan{}, err
+	}
+
+	refreshOpts := &store.RefreshOptions{Scheduled: opts.Flags.IsAutoRefresh}
+	plan, err := refreshCandidates(ctx, st, s.snaps, user, refreshOpts, opts)
+	if err != nil {
+		return updatePlan{}, err
+	}
+
+	return plan, nil
+}
+
+func validateAndInitStoreUpdates(st *state.State, updates map[string]StoreUpdate, opts Options) error {
+	snapstates, err := All(st)
```

Comment:
there were two parts to my comment, the other question is whether we are using the same allSnaps before and after a store call and whether that usage is robust/makes sense, in case theoretically a snap went away while we unlocked to talk to the store

Reply:
Yes, we are using the same set of all snaps during the entire operation. I know we have a test that covers this for the install case (I broke it at some point), but I can't remember what it is called. I'll look around to see if we have one for the update case.

Reply:
Added tests for this in b559572494fd323bf9f92c309f9ad1e5e409fe3d and 762d33a957530e0115f7deb34996190016848e22.